ls1intum/Artemis
# Enable compass for all diagram types

### Important 

- If you give similar elements in the same diagram different assessments with different points, there will be no suggestion for the next one because Compass cannot select which one to suggest.
- There is a scheduled operation that enables automatic feedback suggestions which build clusters to use for categorizing similar elements. Therefore, the exercise should have a due date that is after all the submissions and also must be in the future when you set it otherwise scheduled operation won't be called and you won't see any suggestions.

### Checklist

### Motivation and Context
Currently Compass is only working for Class and Activity Diagrams, however it should be able to work for all diagram types.

### Description
This PR fixes the issues about parsing and similarity checks for other diagram types and enables Compass for all diagram types.

### Steps for Testing
Repeat following steps for all diagram types.

1. Log in to Artemis
2. Navigate to Course Management
3. Create an exercise with automatic assessment suggestions enabled
4. Submit a submission
5. Log in as a student and submit a similar submission
6. Log in as instructor or tutor (Someone who can assess the submissions)
7. Assess one submission
8. Artemis should suggest feedback for the similar elements in other submission

### Screenshots
The following steps must be repeated for all diagram types.
1. Create submission and submit

https://user-images.githubusercontent.com/16825065/122583624-c0c90b80-d059-11eb-9fcd-f7d773e2bf85.mp4

2. Create submission with another student

https://user-images.githubusercontent.com/16825065/122584882-2a95e500-d05b-11eb-8635-ab1c161a7b9a.mp4

3. Set due date of exercise to a future date and wait for until at least 1 minute past that date (Set it to 17:29 at 17:27 and waited until 17:30 for example)

https://user-images.githubusercontent.com/16825065/122585213-8a8c8b80-d05b-11eb-847c-210ec16b6072.mp4

4. Assess the submissions and see the suggestions at the second submission

https://user-images.githubusercontent.com/16825065/122585220-8ceee580-d05b-11eb-989e-f9cff66d2f36.mp4

#### Class Diagram
![image](https://user-images.githubusercontent.com/16825065/122586321-cf64f200-d05c-11eb-8ce0-a994d0fc6657.png)

#### Activity Diagram
![image](https://user-images.githubusercontent.com/16825065/122586587-27035d80-d05d-11eb-9af3-221a9a037062.png)

#### Object Diagram
![image](https://user-images.githubusercontent.com/16825065/122587310-f7088a00-d05d-11eb-8760-70a8755986fd.png)

#### Use Case Diagram
![image](https://user-images.githubusercontent.com/16825065/122586867-747fca80-d05d-11eb-920a-1593b420ad26.png)

#### Communication Diagram
![image](https://user-images.githubusercontent.com/16825065/122587231-df310600-d05d-11eb-9602-8f8b2d588f35.png)

#### Component Diagram
![image](https://user-images.githubusercontent.com/16825065/122587566-4353ca00-d05e-11eb-9beb-74c1f895430a.png)

#### Deployment Diagram
![image](https://user-images.githubusercontent.com/16825065/122588204-1653e700-d05f-11eb-8174-4556e1d83473.png)

#### Petri Net
![image](https://user-images.githubusercontent.com/16825065/122588479-7185d980-d05f-11eb-8c39-a151a1547378.png)

#### Syntax Tree
![image](https://user-images.githubusercontent.com/16825065/122588697-adb93a00-d05f-11eb-86be-4ad783869b19.png)

#### Flowchart
![image](https://user-images.githubusercontent.com/16825065/122588914-efe27b80-d05f-11eb-989c-5f566aa1f555.png)


### Test Coverage
ActivityDiagramParser.java   97%
ClassDiagramParser.java     97%
CommunicationDiagramParser.java  95%
ComponentDiagramParser.java   96%
DeploymentDiagramParser.java   100%
FlowchartParser.java  98%
ObjectDiagramParser.java   96%
PetriNetParser.java  97%
SyntaxTreeParser.java  96%
UMLModelParser.java  95%
UseCaseDiagramParser.java 96%
